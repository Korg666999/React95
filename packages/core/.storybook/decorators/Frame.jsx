import React from 'react';
import '../../components/GlobalStyle/GlobalStyle.css';
import * as themes from '../../components/themes/all';

const Frame = (Story, { globals }) => {
  const { selectedTheme } = globals;

  return (
    <div
      style={{ padding: 10 }}
      className={themes[selectedTheme] || themes.win95}
    >
      <Story />
    </div>
  );
};

export default Frame;
