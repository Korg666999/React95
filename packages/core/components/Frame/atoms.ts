import { sprinkles, Sprinkles } from './sprinkles.css';

function pick<T, K extends keyof T>(obj: T, keys: K[]): Pick<T, K> {
  return keys.reduce(
    (acc, key) =>
      Object.assign(acc, {
        [key]: obj[key],
      }),
    {} as Pick<T, K>,
  );
}

const omit = <T extends object, K extends keyof T>(
  obj: T,
  keys: K[],
): Omit<T, K> => {
  const _ = { ...obj };

  keys.forEach(key => delete _[key]);

  return _;
};

export interface Atoms extends Sprinkles {
  className?: string | string[];
}

const keys = Array.from(sprinkles.properties.keys());
export const extractAtoms = <P extends Sprinkles>(props: P) => [
  pick(props, keys),
  omit(props, keys),
];

export function atoms(atoms: Atoms) {
  const { className, ...rest } = atoms;
  const sprinklesClassNames = sprinkles(rest);

  return [sprinklesClassNames, className].filter(Boolean).join(' ');
}
