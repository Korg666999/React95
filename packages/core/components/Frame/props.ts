import { contract } from '../ThemeProvider/themes/contract.css';

export const positioning = {
  position: true,
  top: true,
  right: true,
} as const;

export const displayAndBoxModel = {
  display: true,
  overflow: true,
  boxSizing: true,

  width: contract.space,
  maxWidth: contract.space,
  minWidth: contract.space,
  height: contract.space,
  maxHeight: contract.space,
  minHeight: contract.space,

  margin: contract.space,
  marginTop: contract.space,
  marginLeft: contract.space,
  marginRight: contract.space,
  marginBottom: contract.space,

  padding: contract.space,
  paddingTop: contract.space,
  paddingLeft: contract.space,
  paddingRight: contract.space,
  paddingBottom: contract.space,
} as const;
