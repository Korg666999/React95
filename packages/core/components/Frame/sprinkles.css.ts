import { createRainbowSprinkles, defineProperties } from 'rainbow-sprinkles';
import {
  background,
  borderRadius,
  borders,
  colors,
  displayAndBoxModel,
  font,
  outline,
  positioning,
  shadows,
  text,
  zIndices,
} from './props';

const displayAndBoxModelAtomicProps = defineProperties({
  dynamicProperties: displayAndBoxModel,
  shorthands: {
    size: ['height', 'width'],
    h: ['height'],
    w: ['width'],
    minH: ['minHeight'],
    minW: ['minWidth'],

    m: ['margin'],
    mr: ['marginRight'],
    ml: ['marginLeft'],
    mt: ['marginTop'],
    mb: ['marginBottom'],
    marginX: ['marginLeft', 'marginRight'],
    marginY: ['marginTop', 'marginBottom'],
    mx: ['marginLeft', 'marginRight'],
    my: ['marginTop', 'marginBottom'],

    p: ['padding'],
    pr: ['paddingRight'],
    pl: ['paddingLeft'],
    pt: ['paddingTop'],
    pb: ['paddingBottom'],
    paddingX: ['paddingLeft', 'paddingRight'],
    paddingY: ['paddingTop', 'paddingBottom'],
    px: ['paddingLeft', 'paddingRight'],
    py: ['paddingTop', 'paddingBottom'],
  },
});

const positioningAtomicProps = defineProperties({
  dynamicProperties: positioning,
});

const colorsAtomicProps = defineProperties({
  dynamicProperties: colors,
});

const backgroundAtomicProps = defineProperties({
  dynamicProperties: background,
  shorthands: {
    bg: ['background'],
    bgColor: ['backgroundColor'],
  },
});

const bordersAtomicProps = defineProperties({
  dynamicProperties: borders,
});

const borderRadiusAtomicProps = defineProperties({
  dynamicProperties: borderRadius,
});

const shadowsAtomicProps = defineProperties({
  dynamicProperties: shadows,
});

const outlineAtomicProps = defineProperties({
  dynamicProperties: outline,
});

const fontAtomicProps = defineProperties({
  dynamicProperties: font,
});

const textAtomicProps = defineProperties({
  dynamicProperties: text,
});

const zIndicesAtomicProps = defineProperties({
  dynamicProperties: zIndices,
});

export const sprinkles = createRainbowSprinkles(
  displayAndBoxModelAtomicProps,
  positioningAtomicProps,
  colorsAtomicProps,
  backgroundAtomicProps,
  bordersAtomicProps,
  borderRadiusAtomicProps,
  shadowsAtomicProps,
  outlineAtomicProps,
  fontAtomicProps,
  textAtomicProps,
  zIndicesAtomicProps,
);

export type Sprinkles = Parameters<typeof sprinkles>[0];
