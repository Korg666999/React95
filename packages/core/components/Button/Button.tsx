import React, { FC } from 'react';

import { button } from './Button.css';
import Frame from '../Frame';
import { FrameProps } from '../Frame/Frame';

export type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> &
  FrameProps;

export const Button: FC<ButtonProps> = props => {
  return (
    <Frame
      as="button"
      {...props}
      className={[button, props.className].join(' ')}
    />
  );
};
