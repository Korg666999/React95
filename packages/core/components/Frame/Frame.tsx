import React, {
  ComponentPropsWithRef,
  ElementType,
  ForwardedRef,
  forwardRef,
} from 'react';
import { atoms, extractAtoms } from './atoms';
import { Sprinkles } from './sprinkles.css';

type FixedForwardRef = <T, P = object>(
  render: (props: P, ref: React.Ref<T>) => React.ReactNode,
) => (props: P & React.RefAttributes<T>) => React.ReactNode;

const fixedForwardRef = forwardRef as FixedForwardRef;

type DistributiveOmit<T, TOmitted extends PropertyKey> = T extends any
  ? Omit<T, TOmitted>
  : never;

const FrameComponent = <TAs extends ElementType>(
  props: {
    as?: TAs;
  } & DistributiveOmit<
    ComponentPropsWithRef<ElementType extends TAs ? 'div' : TAs>,
    'as'
  > &
    Sprinkles,
  ref: ForwardedRef<any>,
) => {
  const { as: Component = 'div', ...other } = props;

  const [atomsProps, propsToForward] = extractAtoms(other as Sprinkles);

  const className = atoms({
    ...atomsProps,
  });

  return <Component {...propsToForward} className={className} ref={ref} />;
};

export const Frame = fixedForwardRef(FrameComponent);
