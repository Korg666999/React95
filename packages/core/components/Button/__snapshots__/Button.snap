// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Button /> > Snapshots > should match snapshot 1`] = `
<div>
  <button
    class="Button_styles__1bzq74l0 "
  />
</div>
`;
